core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortPipelineAggregationBuilder.java

```
@@ -0,0 +1,181 @@
+    public static final String NAME = "bucket_sort";
+
+    private static final ParseField FROM = new ParseField("from");
+    private static final ParseField SIZE = new ParseField("size");
+
+    public static final ConstructingObjectParser<BucketSortPipelineAggregationBuilder, String> PARSER = new ConstructingObjectParser<>(NAME,
+            false, (a, context) -> new BucketSortPipelineAggregationBuilder(context, (List<FieldSortBuilder>) a[0]));
+
+    static {
+        PARSER.declareField(optionalConstructorArg(), (p, c) -> {
+                    List<SortBuilder<?>> sorts = SortBuilder.fromXContent(p);
+                        List<FieldSortBuilder> fieldSorts = new ArrayList<>(sorts.size());
+                        for (SortBuilder<?> sort : sorts) {
+                            if (sort instanceof FieldSortBuilder == false) {
+                                throw new IllegalArgumentException("[" + NAME + "] only supports field based sorting; incompatible sort: ["
+                                        + sort + "]");
+                            }
+                            fieldSorts.add((FieldSortBuilder) sort);
+                        }
+                    return fieldSorts;
+                }, SearchSourceBuilder.SORT_FIELD,
+                ObjectParser.ValueType.OBJECT_ARRAY);
+        PARSER.declareInt(BucketSortPipelineAggregationBuilder::from, FROM);
+        PARSER.declareInt(BucketSortPipelineAggregationBuilder::size, SIZE);
+        PARSER.declareField(BucketSortPipelineAggregationBuilder::gapPolicy, p -> {
+            if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
+                return GapPolicy.parse(p.text().toLowerCase(Locale.ROOT), p.getTokenLocation());
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        }, GAP_POLICY, ObjectParser.ValueType.STRING);
+    }
+
```

Do we need some validation (after parsing) that one of `size` or `sort` is defined?  E.g. in case a user does:

```
{
    "bucket_sort": {
        "sort": []
    }
}
```

---

Hm, not sure. If the user specifies neither, it will just be a no-op. What is the consistent thing to do here?

---

Hrm, not sure either.  I think I'd prefer it to throw an exception, but I could see the alternative argument too.  E.g. your application assembles the queries and it omits both parameters in some situations, so you'd prefer it to not throw exceptions.  Don't have a strong opinion on this one.

@colings86 any thoughts?  Dimitris and I chatted about this on zoom and thought we'd see what you had to say :)

---

I think we should throw an exception. It seems like a clear mistake if the user does this and I'm not sure it would actually be a no-op because it would output a complete copy of the target aggregation in the response as well as the original?

---

It would be no-op because this is sorting inline.

---

ah yes, sorry. I'd still like to throw an exception though as this is almost certainly a mistake by the user so they should be notified so they can correct it